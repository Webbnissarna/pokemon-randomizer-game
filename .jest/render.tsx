import { render as rtlRender } from "@testing-library/react";
import { InitializeColorMode, ThemeProvider } from "theme-ui";

import theme from "../utils/theme";

import "./style.css";

export function render(ui: React.ReactElement) {
  return rtlRender(
    <>
      <InitializeColorMode />
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </>
  );
}
