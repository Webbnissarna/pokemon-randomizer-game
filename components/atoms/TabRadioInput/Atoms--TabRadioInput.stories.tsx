import React, { useState } from "react";
import { Flex } from "theme-ui";
import TabRadioInput from "./TabRadioInput";
import { OnChangeEvent } from "./TabRadioInput.types";

export function Default() {
  const values = ["Solo", "2-player"];
  const [selected, setSelected] = useState(values[0]);

  const onChange = (e: OnChangeEvent) => {
    setSelected(e.target.value);
  };
  return (
    <Flex sx={{ flexDirection: "column", gap: 20 }}>
      <TabRadioInput onChange={onChange} values={values} selected={selected} />
      <span>Selected: {selected}</span>
    </Flex>
  );
}
