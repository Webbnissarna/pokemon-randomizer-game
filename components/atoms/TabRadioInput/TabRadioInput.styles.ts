import { ThemeUIStyleObject } from "theme-ui";

export const containerStyle: ThemeUIStyleObject = {
  "& > div:first-of-type label": {
    borderTopLeftRadius: "md",
    borderBottomLeftRadius: "md",
  },
  "& > div:last-of-type label": {
    borderTopRightRadius: "md",
    borderBottomRightRadius: "md",
  },
};

export const valueRootStyle: ThemeUIStyleObject = {
  flexGrow: 1,
  display: "flex",
  justifyContent: "center",
  alignItems: "center",

  "& input": {
    display: "none",
  },

  "& input:checked ~ label": {
    color: "secondaryText",
    backgroundColor: "secondary",
    filter: "brightness(0.6)",
    opacity: 0.8,
    cursor: "unset",
  },

  "& input:not(:checked) ~ label": {
    "&:hover": {
      filter: "brightness(1.15)",
    },
    "&:active": {
      filter: "brightness(0.95)",
    },
  },
};

export const labelStyle: ThemeUIStyleObject = {
  display: "flex",
  justifyContent: "center",
  alignItems: "center",
  color: "neutralText",
  backgroundColor: "neutral",
  height: "sm",
  fontFamily: "inherit",
  fontSize: "md",
  boxShadow: "md",
  cursor: "pointer",

  transition:
    "filter 0.1s ease-out, background 0.1s ease-out, color 0.1s ease-out",
};
