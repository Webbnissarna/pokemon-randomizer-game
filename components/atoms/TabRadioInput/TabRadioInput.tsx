import React, { useId } from "react";
import { Box, Flex, Label } from "theme-ui";

interface TabRadioInputProps {
  values: string[];
  selected: string;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

export default function TabRadioInput({
  values,
  selected,
  onChange,
}: TabRadioInputProps): JSX.Element {
  const id = useId();

  return (
    <Flex
      sx={{
        "& > div:first-of-type label": {
          borderTopLeftRadius: "md",
          borderBottomLeftRadius: "md",
        },
        "& > div:last-of-type label": {
          borderTopRightRadius: "md",
          borderBottomRightRadius: "md",
        },
      }}
    >
      {values.map((value) => (
        <Box
          key={value}
          sx={{
            flexGrow: 1,
            display: "flex",
            justifyContent: "center",
            alignItems: "center",

            "& input": {
              display: "none",
            },

            "& input:checked ~ label": {
              color: "secondaryText",
              backgroundColor: "secondary",
              filter: "brightness(0.6)",
              opacity: 0.8,
              cursor: "unset",
            },

            "& input:not(:checked) ~ label": {
              "&:hover": {
                filter: "brightness(1.15)",
              },
              "&:active": {
                filter: "brightness(0.95)",
              },
            },
          }}
        >
          <input
            name={id}
            type="radio"
            id={`${id}-${value}`}
            checked={value === selected}
            value={value}
            onChange={onChange}
          />
          <Label
            htmlFor={`${id}-${value}`}
            sx={{
              display: "flex",
              justifyContent: "center",
              alignItems: "center",
              color: "neutralText",
              backgroundColor: "neutral",
              height: 30,
              fontFamily: "inherit",
              fontSize: "md",
              boxShadow: "md",
              cursor: "pointer",

              transition:
                "filter 0.1s ease-out, background 0.1s ease-out, color 0.1s ease-out",
            }}
          >
            {value}
          </Label>
        </Box>
      ))}
    </Flex>
  );
}
