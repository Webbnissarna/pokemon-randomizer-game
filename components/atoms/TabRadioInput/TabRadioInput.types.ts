export type OnChangeEvent = React.ChangeEvent<HTMLInputElement>;

export interface TabRadioInputProps {
  values: string[];
  selected: string;
  onChange: (event: OnChangeEvent) => void;
}
