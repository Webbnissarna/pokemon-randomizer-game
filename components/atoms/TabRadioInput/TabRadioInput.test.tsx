import { screen } from "@testing-library/react";
import { render } from "../../../.jest/render";
import userEvent from "@testing-library/user-event";
import TabRadioInput from ".";

describe("Tab Radio Input", () => {
  it("renders", () => {
    const func = () =>
      render(
        <TabRadioInput
          values={["test"]}
          selected="test"
          onChange={() => undefined}
        />
      );

    expect(func).not.toThrow();
  });

  it("shows values", () => {
    const values = ["Solo", "2-player", "vs bot"];

    render(
      <TabRadioInput
        values={values}
        selected={values[0]}
        onChange={() => undefined}
      />
    );

    values.map((value) => screen.getByLabelText(value));

    expect(true).toBeTruthy();
  });

  it("marks selected", () => {
    const values = ["Solo", "2-player", "vs bot"];
    const selected = values[0];

    render(
      <TabRadioInput
        values={values}
        selected={selected}
        onChange={() => undefined}
      />
    );
    const element = screen.getByLabelText<HTMLInputElement>(selected);

    expect(element.checked).toBeTruthy();
  });

  it("calls onChange", async () => {
    const values = ["Solo", "2-player", "vs bot"];
    const valueToClick = values[1];
    const callback = jest.fn();
    const user = userEvent.setup();

    render(
      <TabRadioInput values={values} selected={values[0]} onChange={callback} />
    );
    const element = screen.getByLabelText<HTMLInputElement>(valueToClick);
    await user.click(element);

    expect(callback).toHaveBeenCalledWith(
      expect.objectContaining({
        target: expect.objectContaining({
          value: valueToClick,
        }),
      })
    );
  });
});
