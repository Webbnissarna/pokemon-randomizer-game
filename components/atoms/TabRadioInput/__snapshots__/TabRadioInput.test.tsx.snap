// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab Radio Input renders 1`] = `
<DocumentFragment>
  <script>
    (function() { try {
  var mode = localStorage.getItem('theme-ui-color-mode');
  if (!mode) return
  document.documentElement.classList.add('theme-ui-' + mode);
} catch (e) {} })();
  </script>
  <div
    class="css-1qwd1qz-Box"
  >
    <div
      class="css-1d2fn-Box"
    >
      <input
        checked=""
        id=":r0:-test"
        name=":r0:"
        type="radio"
        value="test"
      />
      <label
        class="css-dzydor-Box"
        for=":r0:-test"
      >
        test
      </label>
    </div>
  </div>
</DocumentFragment>
`;
